apps/smart-contracts/token/test/MiniSales.test.ts

```diff
@@ -80,9 +89,246 @@ describe('=> MiniSales', () => {
     })
   })
 
-  describe('# setPrice', () => {
+  describe('# purchase', () => {
+    let contractSaleTokenBalance: BigNumber
+    let purchaseAmount: BigNumber
+    let paymentNeeded: BigNumber
+    let paymentAmount: BigNumber
+    let purchaser: SignerWithAddress
+    let recipient: SignerWithAddress
+
     beforeEach(async () => {
       await setupMiniSales()
+      contractSaleTokenBalance = await saleToken.balanceOf(miniSales.address)
+      purchaseAmount = contractSaleTokenBalance
+      paymentNeeded = purchaseAmount.mul(testPrice).div(testDenominator)
+      paymentAmount = paymentNeeded
+      purchaser = user1
+      recipient = user1
+    })
+
+    it("reverts if price doesn't match", async () => {
+      const currentPrice = await miniSales.getPrice()
+
+      await expect(
+        miniSales
+          .connect(purchaser)
+          .purchase(recipient.address, purchaseAmount, currentPrice.add(1))
+      ).revertedWith('Price does not match')
+    })
+
+    it('reverts if purchase hook reverts', async () => {
+      expect(await miniSales.getPrice()).to.eq(testPrice)
+      fakePurchaseHook.hook.reverts()
+
+      await expect(
+        miniSales.connect(purchaser).purchase(recipient.address, purchaseAmount, testPrice)
+      ).to.be.reverted
+    })
+
+    it('reverts if insufficient sale token in contract', async () => {
+      purchaseAmount = contractSaleTokenBalance.add(1)
+      paymentNeeded = purchaseAmount.mul(testPrice).div(testDenominator)
+      paymentAmount = paymentNeeded
+      await paymentToken.connect(owner).transfer(purchaser.address, paymentAmount)
+      await paymentToken.connect(purchaser).approve(miniSales.address, paymentAmount)
+
+      await expect(
+        miniSales.connect(purchaser).purchase(recipient.address, purchaseAmount, testPrice)
+      ).to.revertedWith('ERC20: transfer amount exceeds balance')
+    })
+
+    it('reverts if insufficient payment token provided', async () => {
+      paymentAmount = paymentNeeded.sub(1)
+      await paymentToken.connect(owner).transfer(purchaser.address, paymentAmount)
```

Comment:
clearer just inputting directly
```suggestion
      await paymentToken.connect(owner).transfer(purchaser.address, paymentNeeded.sub(1))
```